File: quiz-api/src/quiz_api/routes/chapters.py
"""Chapter Management Routes."""

from http import (
    HTTPMethod,
    HTTPStatus,
)

from flask import (
    Blueprint,
    jsonify,
    request,
)
from flask_jwt_extended import (
    get_jwt_identity,
    jwt_required,
)

from quiz_api.models.database import db
from quiz_api.models.models import (
    Chapter,
    Subject,
    User,
)
from quiz_api.models.schemas import (
    ChapterSchema,
    ChapterUpdateSchema,
    SearchSchema,
)
from quiz_api.utils.search import search_chapters

# Define Blueprint
chapters_bp = Blueprint("chapters", __name__)


@chapters_bp.route("/subjects/<int:subject_id>/chapters", methods=[HTTPMethod.POST])
@jwt_required()
def create_chapter(subject_id: int):
    """Create a new chapter (Admin only)."""
    try:
        # Check if user is admin
        current_user_id = int(get_jwt_identity())
        current_user: User | None = db.session.get(User, current_user_id)
        if not current_user or current_user.role != "admin":
            return jsonify({"message": "Unauthorized"}), HTTPStatus.FORBIDDEN

        # Check if subject exists
        subject: Subject | None = db.session.get(Subject, subject_id)
        if not subject:
            return jsonify({"message": "Subject not found"}), HTTPStatus.NOT_FOUND

        # Validate request data
        chapter_data = ChapterSchema(**request.get_json())

        # Create new chapter
        new_chapter = Chapter(name=chapter_data.name, description=chapter_data.description, subject_id=subject_id)

        db.session.add(new_chapter)
        db.session.commit()

        return (
            jsonify(
                {
                    "message": "Chapter created successfully",
                    "chapter": {
                        "id": new_chapter.id,
                        "name": new_chapter.name,
                        "description": new_chapter.description,
                        "subject_id": new_chapter.subject_id,
                    },
                }
            ),
            HTTPStatus.CREATED,
        )

    except ValueError as e:
        return jsonify({"message": str(e)}), HTTPStatus.BAD_REQUEST
    except Exception as e:
        raise
    finally:
        db.session.close()


@chapters_bp.route("/subjects/<int:subject_id>/chapters", methods=[HTTPMethod.GET])
def get_subject_chapters(subject_id: int):
    """Get all chapters under a subject."""
    try:
        # Check if subject exists
        subject: Subject | None = db.session.get(Subject, subject_id)
        if not subject:
            return jsonify({"message": "Subject not found"}), HTTPStatus.NOT_FOUND

        chapters = Chapter.query.filter_by(subject_id=subject_id).all()
        chapters_list = [
            {
                "id": chapter.id,
                "name": chapter.name,
                "description": chapter.description,
                "subject_id": chapter.subject_id,
                "created_at": chapter.created_at.isoformat(),
                "updated_at": chapter.updated_at.isoformat() if chapter.updated_at else None,
            }
            for chapter in chapters
        ]

        return jsonify(chapters_list), HTTPStatus.OK
    except Exception as e:
        raise
    finally:
        db.session.close()


@chapters_bp.route("/chapters/<int:chapter_id>", methods=[HTTPMethod.GET])
def get_chapter(chapter_id: int):
    """Get details of a specific chapter."""
    try:
        chapter: Chapter | None = db.session.get(Chapter, chapter_id)
        if not chapter:
            return jsonify({"message": "Chapter not found"}), HTTPStatus.NOT_FOUND

        response = {
            "id": chapter.id,
            "name": chapter.name,
            "description": chapter.description,
            "subject_id": chapter.subject_id,
            "created_at": chapter.created_at.isoformat(),
            "updated_at": chapter.updated_at.isoformat() if chapter.updated_at else None,
        }
        return jsonify(response), HTTPStatus.OK
    except Exception as e:
        raise
    finally:
        db.session.close()


@chapters_bp.route("/chapters/<int:chapter_id>", methods=[HTTPMethod.PATCH])
@jwt_required()
def update_chapter(chapter_id: int):
    """Update a chapter's details (Admin only)."""
    try:
        # Check if user is admin
        current_user_id = int(get_jwt_identity())
        current_user: User | None = db.session.get(User, current_user_id)
        if not current_user or current_user.role != "admin":
            return jsonify({"message": "Unauthorized"}), HTTPStatus.FORBIDDEN

        chapter: Chapter | None = db.session.get(Chapter, chapter_id)
        if not chapter:
            return jsonify({"message": "Chapter not found"}), HTTPStatus.NOT_FOUND

        # Validate update data
        update_data = ChapterUpdateSchema(**request.get_json())

        # Update fields if provided
        if update_data.name:
            chapter.name = update_data.name
        if update_data.description:
            chapter.description = update_data.description

        db.session.commit()
        return jsonify({"message": "Chapter updated successfully"}), HTTPStatus.OK
    except ValueError as e:
        return jsonify({"message": str(e)}), HTTPStatus.BAD_REQUEST
    except Exception as e:
        raise
    finally:
        db.session.close()


@chapters_bp.route("/chapters/<int:chapter_id>", methods=[HTTPMethod.DELETE])
@jwt_required()
def delete_chapter(chapter_id: int):
    """Delete a chapter (Admin only)."""
    try:
        # Check if user is admin
        current_user_id = int(get_jwt_identity())
        current_user: User | None = db.session.get(User, current_user_id)
        if not current_user or current_user.role != "admin":
            return jsonify({"message": "Unauthorized"}), HTTPStatus.FORBIDDEN

        chapter: Chapter | None = db.session.get(Chapter, chapter_id)
        if not chapter:
            return jsonify({"message": "Chapter not found"}), HTTPStatus.NOT_FOUND

        db.session.delete(chapter)
        db.session.commit()
        return jsonify({"message": "Chapter deleted successfully"}), HTTPStatus.OK
    except Exception as e:
        raise
    finally:
        db.session.close()


@chapters_bp.route("/subjects/<int:subject_id>/chapters/search", methods=[HTTPMethod.GET])
def search_subject_chapters(subject_id: int):
    """Search chapters within a subject using Full-Text Search."""
    try:
        search_params = SearchSchema(**request.args)
        query = search_params.q

        # Check if subject exists
        subject: Subject | None = db.session.get(Subject, subject_id)
        if not subject:
            return jsonify({"message": "Subject not found"}), HTTPStatus.NOT_FOUND

        if not query:
            # Return all chapters for this subject if no query
            chapters = (
                Chapter.query.filter_by(subject_id=subject_id)
                .limit(search_params.limit)
                .offset(search_params.offset)
                .all()
            )
            chapters_list = [
                {
                    "id": chapter.id,
                    "name": chapter.name,
                    "description": chapter.description,
                    "subject_id": chapter.subject_id,
                    "created_at": chapter.created_at.isoformat(),
                    "updated_at": chapter.updated_at.isoformat() if chapter.updated_at else None,
                }
                for chapter in chapters
            ]
        else:
            # Use FTS to search chapters
            results = search_chapters(query, limit=search_params.limit, offset=search_params.offset)

            # Filter results to only include chapters from this subject
            chapters_list = [
                {
                    "id": row[0],
                    "name": row[1],
                    "description": row[2],
                    "subject_id": row[3],
                    "created_at": row[4] if isinstance(row[4], str) else row[4].isoformat() if row[4] else None,
                    "updated_at": row[5] if isinstance(row[5], str) else row[5].isoformat() if row[5] else None,
                }
                for row in results
                if row[3] == subject_id
            ]

        # Return with metadata
        response = {
            "items": chapters_list,
            "total": len(chapters_list),
            "limit": search_params.limit,
            "offset": search_params.offset,
        }

        return jsonify(response), HTTPStatus.OK
    except Exception as e:
        raise
    finally:
        db.session.close()

File: quiz-web/js/api/chapters.js
const API_BASE_URL = config.API_URL;

// Import handleApiResponse from utils
import { handleApiResponse } from '/js/utils.js';

/**
 * Get all chapters for a specific subject
 * @param {number} subjectId - The ID of the subject
 * @returns {Promise<Array>} - Array of chapter objects
 */
export async function getChapters(subjectId) {
    try {
        const response = await fetch(`${API_BASE_URL}/subjects/${subjectId}/chapters`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('token')}`
            }
        });

        return await handleApiResponse(response);
    } catch (error) {
        console.error('Error fetching chapters:', error);
        throw error;
    }
}

/**
 * Get a specific chapter by ID
 * @param {number} chapterId - The ID of the chapter
 * @returns {Promise<Object>} - Chapter object
 */
export async function getChapter(chapterId) {
    try {
        const response = await fetch(`${API_BASE_URL}/chapters/${chapterId}`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('token')}`
            }
        });

        return await handleApiResponse(response);
    } catch (error) {
        console.error('Error fetching chapter:', error);
        throw error;
    }
}

/**
 * Create a new chapter
 * @param {number} subjectId - The ID of the subject
 * @param {Object} data - Chapter data (name, description)
 * @returns {Promise<Object>} - Created chapter object
 */
export async function createChapter(subjectId, data) {
    try {
        const response = await fetch(`${API_BASE_URL}/subjects/${subjectId}/chapters`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('token')}`
            },
            body: JSON.stringify(data)
        });

        return await handleApiResponse(response);
    } catch (error) {
        console.error('Error creating chapter:', error);
        throw error;
    }
}

/**
 * Update an existing chapter
 * @param {number} chapterId - The ID of the chapter to update
 * @param {Object} data - Updated chapter data
 * @returns {Promise<Object>} - Updated chapter object
 */
export async function updateChapter(chapterId, data) {
    try {
        const response = await fetch(`${API_BASE_URL}/chapters/${chapterId}`, {
            method: 'PATCH',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('token')}`
            },
            body: JSON.stringify(data)
        });

        return await handleApiResponse(response);
    } catch (error) {
        console.error('Error updating chapter:', error);
        throw error;
    }
}

/**
 * Delete a chapter
 * @param {number} chapterId - The ID of the chapter to delete
 * @returns {Promise<boolean>} - True if successful
 */
export async function deleteChapter(chapterId) {
    try {
        const response = await fetch(`${API_BASE_URL}/chapters/${chapterId}`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('token')}`
            }
        });

        await handleApiResponse(response);
        return true;
    } catch (error) {
        console.error('Error deleting chapter:', error);
        throw error;
    }
}

/**
 * Search chapters within a subject
 * @param {number} subjectId - The ID of the subject
 * @param {string} query - Search query
 * @param {number} limit - Maximum number of results (default: 10)
 * @param {number} offset - Number of results to skip (default: 0)
 * @returns {Promise<Object>} - Search results with metadata
 */
export async function searchChapters(subjectId, query, limit = 10, offset = 0) {
    try {
        const params = new URLSearchParams({
            q: query,
            limit,
            offset
        });

        const response = await fetch(`${API_BASE_URL}/subjects/${subjectId}/chapters/search?${params}`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${localStorage.getItem('token')}`
            }
        });

        return await handleApiResponse(response);
    } catch (error) {
        console.error('Error searching chapters:', error);
        throw error;
    }
}
